### ISSUE TITLE
Upgrade acinerella/ffmpeg version

### ISSUE THREAD

**Issue by lukeIam** [Vocaluxe/Vocaluxe#197]
We are currently using a quite old ffmpeg version, but acinerella was recently updated (and supports now the new ffmpeg version): https://github.com/Vocaluxe/Vocaluxe/issues/162#issuecomment-173605634

As we are using a customized version of acinerella and the source seems to be lost, there will be some rewriting of code required and also some Interfaces must be changed.

**Todo**:
- [x] - check for performance impacts
- [x] -  move some code to the native lib to reduce the number of interop calls
- [x] -  some other small things
- [x] - create nuget packages for new DLLs

Edit:
- [x] - Compare my code with the discovered code in `Vocaluxe/Lib/Video/Acinerella` and merge them
- [x] - Update code at `Vocaluxe/Lib/Video/Acinerella` (needed for Linux build)
- [x] - Check if we can and want to upgrade to the real latest ffmpeg version
- [x] - Merge this to the develop branch

**Comment by lukeIam** [comment]
Here is a version with all videos playing:
https://github.com/lukeIam/Vocaluxe/tree/216375ab10e24493e73d5ac88760985895bd6a15

Use it with:
https://github.com/lukeIam/acinerella/tree/7ac8834b350ada0c650a9029ed8ea371258f2242

Inject the new dlls here:
```
Vocaluxe\packages\vocaluxe.dependencies.acinerella.1.0.1\[x64|x86]
Vocaluxe\packages\vocaluxe.dependencies.ffmpeg.1.0.1\[x64|x86]
```

**Todo**:
- check for performance impacts
- move some code to the native lib to reduce the number of interop calls
- some other small things

**Comment by lukeIam** [comment]
New version:
https://github.com/lukeIam/Vocaluxe/commit/e4deacf0b66666968be1625112acc3d14bfd0883
(acinerella building and DLL injection not longer required)

**Comment by lukeIam** [comment]
I did some performance testing:
- let the startup video play completely
- select a h264 song (which had a huge GAP)
- sing two lines of the song
- select a h264 song (which had a small GAP)
- sing two lines of the song
- quit Vocaluxe

(As all this was done by hand the results where not that precipice.)

Used:
- Diagnostic Tools
- dotTrace

**Results:**
- video preview for the huge-GAP-song started faster (~2/3 of the time)
- CPU usage for decoding decreased ~50%

Disclaimer: this was just a quick check - feel free to do you own tests and report the results.

**Comment by lukeIam** [comment]
Here are the binaries for easy testing:
[newAcinerella build x64](https://ci.appveyor.com/api/buildjobs/uns5p5uiqc8g4dmx/artifacts/Vocaluxe_newAcinerella_Windows_x64.zip)
[newAcinerella build x68](https://ci.appveyor.com/api/buildjobs/lm9xm1s76eby4prh/artifacts/Vocaluxe_newAcinerella_Windows_x86.zip)

**Comment by flokuep** [comment]
Did also some smaller tests and can confirm some improvements :+1:  - Thanks!

**Comment by derivator** [comment]
> As we are using a customized version of acinerella and the source seems to be lost

Is the code in `Vocaluxe/Lib/Video/Acinerella` not the right version? I used that for the linux build and it worked fine...

I'll try my luck with the updated version...
By the way, we should think about adding the [customized version](https://github.com/lukeIam/acinerella/tree/Vocaluxe_Adjustments) to the Vocaluxe repository again, either directly or (preferably, I think) as a git submodule (more complex but also more elegant).
We want this because on Linux, prebuilding acinerella isn't really an option unless we want to statically link all of ffmpeg... I also think it's nicer to have our custom changes in the same place as Vocaluxe, so we won't lose the code or have someone try to build with the wrong version by accident.

If we could get our changes merged in some way and use the official branch that would be great of course.

**Comment by derivator** [comment]
Upgraded version is working on linux, but could we get the dependencies separated by platform? The error conditions won't even let me build unless I have nuget download prebuilt windows only acinerella dlls, which is very annoying in case of a server outage. This also applies to hidapi, gstreamer, ffmpeg, PitchTracker, of course...

**Comment by lukeIam** [comment]
> Is the code in Vocaluxe/Lib/Video/Acinerella not the right version?

:grimacing: That would might has helped writing the functions...
Maybe we should also test if that implementation is even faster then mine (I just tried to get it working).

> By the way, we should think about adding the customized version to the Vocaluxe repository again, either directly or (preferably, I think) as a git submodule (more complex but also more elegant).

I would vote for using the direct files.
Git submodules are not so easy to [handle](https://codingkilledthecat.wordpress.com/2012/04/28/why-your-company-shouldnt-use-git-submodules/)

**Comment by lukeIam** [comment]
> Upgraded version is working on linux, but could we get the dependencies separated by platform? 

As far as I know there is no possibility to have different nuget dependencies in one single solution.

>The error conditions won't even let me build unless I have nuget download prebuilt windows only acinerella dlls, which is very annoying in case of a server outage.

At least under Windows nuget is caching the packages in the (untracked) folder `packages`. So I'm loading every package only once (actually I'm working here with a 3G connection)

What I can do is:
- do not copy the files from `packages` to `lib` folder to keep the `lib` folder cleaner
- add a condition to that error statements - but I'm not sure if then nuget does not recognize them anymore and creates new ones

**Comment by janus-reith** [comment]
Acinerella got another update yesterday which made it compatible to the latest FFMPeg version.

**Comment by lukeIam** [comment]
> What I can do is:
> 
> do not copy the files from packages to lib folder to keep the lib folder cleaner
> add a condition to that error statements - but I'm not sure if then nuget does not recognize them anymore and creates new ones

> This also applies to hidapi, gstreamer, ffmpeg, PitchTracker, of course...

@derivator Please try: https://github.com/lukeIam/Vocaluxe/tree/nugetNoLinuxCopy

I disabled copying the files to the lib folder and disabled the error on missing this packages:

- hidapi
- gstreamer
- ffmpeg
- acinerella
- PitchTracker (this nuget package is currently not used -  windows also builds it from source)

As already told the nuget packed is always loaded on a restore (but normally from cache) and there is no way around this besides creating an other project.

**Comment by lukeIam** [comment]
Updated ToDo in issue:

> - Compare my code with the discovered code in Vocaluxe/Lib/Video/Acinerella and merge them
> - Update code at Vocaluxe/Lib/Video/Acinerella (needed for Linux build)
> - Check if we can and want to upgrade to the real latest ffmpeg version
> - Merge this to the develop branch

**Comment by derivator** [comment]
Thanks, I will try it later.

**Comment by lukeIam** [comment]
Moved the nuget-linux-part to a new issue: #212

**Comment by lukeIam** [comment]
I was comparing the two customized versions and found two things:
- Do we need to change (https://github.com/lukeIam/acinerella/blob/Vocaluxe_Adjustments/src/acinerella.c#L978) from:
```c#
if (ac_skip_package(pckt, pDecoder) != 0)
```
to
```c#
if (pckt->stream_index == pDecoder->stream_index && ac_skip_package(pckt, pDecoder) != 0)
``` 
?

- Do we need something like the ac_sync_video() function? (https://github.com/Vocaluxe/Vocaluxe/blob/develop/Vocaluxe/Lib/Video/Acinerella/acinerella.c#L371)
I'm not really sure what it is doing...

**Comment by basisbit** [comment]
the new accinerella is GPLv3. Doesn't this mean that we also have to change vocaluxe to GPLv3 to be allowed to integrate accinerella code?

**Comment by lukeIam** [comment]
@basisbit As far as I know Vocaluxe is already under GPLv3:
https://github.com/Vocaluxe/Vocaluxe/blob/develop/LICENSE.txt

So there shouldn't be a problem from this side.

**Comment by lukeIam** [comment]
> - Check if we can and want to upgrade to the real latest ffmpeg version

Done! Yes we can 😉 
I modified acinerella to work with the latest ffmpeg 3.2:
https://github.com/lukeIam/acinerella/tree/Vocaluxe_Adjustments2

First tests look good...

**Comment by lukeIam** [comment]
> I was comparing the two customized versions and found two things:
> 
> Do we need to change (https://github.com/lukeIam/acinerella/blob/Vocaluxe_Adjustments/src/acinerella.c#L978) from:
> if (ac_skip_package(pckt, pDecoder) != 0)
> to
> 
> if (pckt->stream_index == pDecoder->stream_index && ac_skip_package(pckt, pDecoder) != 0)
> ?

I thought about it and it makes sense to check `stream_index` -> Added

> Do we need something like the ac_sync_video() function? (https://github.com/Vocaluxe/Vocaluxe/blob/develop/Vocaluxe/Lib/Video/Acinerella/acinerella.c#L371)
> I'm not really sure what it is doing...

I had a deeper look in the code and I'm doing this adjustment in `ac_decode_package()` and `ac_skip_package()` -> nothing to change here

> - Compare my code with the discovered code in Vocaluxe/Lib/Video/Acinerella and merge them
Done (...as far as I can tell)

**Comment by lukeIam** [comment]
Now we need to figure out how stable the newAcinerella change is...

**How to test:**  

Use the [newAcinerella branch](https://github.com/Vocaluxe/Vocaluxe/tree/newAcinerella) (no changes here since the last time)

Use the files from one of those zips:  
[x86_acinerella_Vocaluxe_Adjustments2_27f0984.zip](https://github.com/Vocaluxe/Vocaluxe/files/658193/x86_acinerella_Vocaluxe_Adjustments2_27f0984.zip)
[x64_acinerella_Vocaluxe_Adjustments2_27f0984.zip](https://github.com/Vocaluxe/Vocaluxe/files/658200/x64_acinerella_Vocaluxe_Adjustments2_27f0984.zip)

**Important**:  
Before testing other branches undo this little hacky injection by deleting `\packages\vocaluxe.dependencies.ffmpeg.1.0.2`, `\packages\vocaluxe.dependencies.acinerella.1.0.2` and `\Output\libs\unmanaged`

**Comment by basisbit** [comment]
could you please also do a pull request for your acinerella changes to the original repo? thanks!

**Comment by lukeIam** [comment]
> could you please also do a pull request for your acinerella changes to the original repo? thanks!

For now I'm not sure if it's stable enough - but if you want to use it now, you can use the version from my fork.

**Comment by lukeIam** [comment]
> - Update code at Vocaluxe/Lib/Video/Acinerella (needed for Linux build)

Done

Also created nuget packages for the new version and updated [lukeIam/Vocaluxe:newAcinerella](https://github.com/lukeIam/Vocaluxe/tree/newAcinerella) to use them.

There are also a compiled test-versions:
https://ci.appveyor.com/project/lukeIam/vocaluxe/build/49

**Comment by lukeIam** [comment]
closed by #248
